docs/aqua-book/language/types.md

```diff
@@ -2,23 +2,52 @@
 
 ## Scalars
 
-Scalar types follow the Wasm IT notation.
+Scalar types in Aqua follow the Wasm IT notation.
 
 * Unsigned numbers: `u8`, `u16`, `u32`, `u64`
 * Signed numbers: `i8`, `i16`, `i32`, `i64`
 * Floats: `f32`, `f64`
 * Boolean: `bool`
 * String: `string`
-* Structures (product types): see [Structures](types.md#structures) below
-* Arrays: see [Collection Types](types.md#collection-types) below
 
-## Literals
+## Data Types
 
-You can pass booleans (`true`, `false`), numbers, double-quoted strings as literals.
+In Aqua terminology, data types are types that actually represent value at runtime. In other words, it is any type that is an immutable combination of scalars:
+- [Scalars](#scalars)
+- [Immutable Collections](#immutable-collections)
+- [Structures](#structures)
+
+## Collection Types
+
+There are two kinds of collections in Aqua: immutable and appendable. To denote a collection of a given type `T`, use the quantifier before the type name like so:
+- `?T` for an immutable option of `T`, e.g. `?[]i64`, `?string`
+- `[]T` for an immutable array of `T`, e.g. `[]u32`, `[]?Result`
+- `*T` for an appendable stream of `T`, e.g. `*[]Response`, `*?string`
+
+### Immutable Collections
+
+Aqua has two types of immutable collections: 
+- Arrays containing a fixed number of elements of the same type. The quantifier is `[]`, e.g. `[]string`, `[]?Report`, `[][]i64` are all valid array types. To access an element of an array, use: 
+  - The square brackets notation with any expression of integer type inside, e.g. `arr[10]`, `arr[arr.length - 1]`.
+  - The bang notation, optionally followed by a number or a variable, e.g. `arr!10`, `arr!index`, `arr!` (the same as `arg!0`).
+- Options containing none or one element of a given type. The quantifier is `?`, e.g. `?string`, `?[]Error` are all valid option types. To access an element of an option, use the bang notation, e.g. `opt!` or `opt!0`. There is a literal that represents an empty option: `nil`.
+
+### Appendable Collections
+
+Aqua has appendable [streams](./crdt-streams.md). The quantifier is `*`, e.g. `*?string`, `*Result`, `*[]i64` are all valid stream types. To access an element of a stream, use:
+  - The square brackets notation, with any expression of integer type inside, e.g. `stream[10]`, `stream[requests.length - 1]`
```

Comment:
I would mention somewhere that this indexed access is locally deterministic, but only locally

Reply:
Could you elaborate on what do you mean by "locally" in this context?